syntax = "proto3";

package server;

service Server {
    rpc find_successor(FindSucRequest) returns (FindSucResponse) {}
    rpc live_predecessor(PredecessorRequest) returns (PredecessorResponse) {}
    rpc find_succlist(FindSucclistRequest) returns (FindSucclistResponse) {}
    rpc find_predecessor(EmptyRequest) returns (FindPredResponse) {}
    rpc rectify(RectifyRequest) returns (RectifyResponse) {}
    rpc get_node_status(EmptyRequest) returns (NodeStatus) {}
    //defined by ourselves
    rpc replicate_entries(ReplicateRequest) returns (ReplicateResponse){}
    rpc get(GetRequest) returns (GetResponse) {}
    rpc put(PutRequest) returns (PutResponse) {}
}

enum ReturnCode {
    SUCCESS = 0;
    FAILURE = 1;
}

message NodeStatus {
    int32 id = 1;
    string ip = 2;
    int32 pred_id = 3;
    string pred_ip = 4;
    repeated int32 suclist_id = 5;
    repeated string suclist_ip = 6;
    repeated int32 finger_id = 7;
    repeated string finger_ip = 8;
}

message EmptyRequest {
}

message RectifyResponse {
    ReturnCode ret = 1;
}

//Not needed to include inc id and ip
message RectifyRequest {
    int32 id = 1;
    string ip = 2;
}

message FindSucRequest {
    int32 id = 1;
    int32 inc_id = 2;
    string inc_ip = 3;
}

message FindSucResponse {
    int32 id = 1;
    string ip = 2;
}

message FindPredResponse {
    int32 id = 1;
    string ip = 2;
}

message FindSucclistRequest{
    int32 inc_id = 1;
    string inc_ip = 2;
}

message FindSucclistResponse {
    repeated int32 id_list = 1;
    repeated string ip_list = 2;
}

//Not needed to insert inc_id and ip?
message PredecessorRequest {
    int32 id = 1;
}

message PredecessorResponse {
    ReturnCode ret = 1;
}

//hash result of the key
message LogEntry {
    int32 hashID = 1;
    string key = 2;
    string val = 3;
}

message ReplicateRequest{
    repeated LogEntry entries = 1;
}

message ReplicateResponse{
    ReturnCode ret = 1;
}

message GetRequest {
    string key = 1;
}

message GetResponse {
    ReturnCode ret = 1;
    string response = 2;
    int32 nodeID = 3;
    string nodeIP = 4;
}

message PutRequest {
    string key = 1;
    string value = 2;
}

message PutResponse {
    ReturnCode ret = 1;
    int32 nodeID = 2;
    string nodeIP = 3;
}

